### PR TITLE
add OpenDose extension

### PR DESCRIPTION
**Pull request by BishopWolf**
- [x] Extension has a reasonable name (not too general, not too narrow, suggests what the extension is for)
- [x] Repository name is Slicer+ExtensionName
- [x] Repository is associated with `3d-slicer-extension` GitHub topic so that it is listed [here](https://github.com/topics/3d-slicer-extension). To edit topics, click "Manage topics" on your repository's main page, on the left, near the top (just below the repository description). To learn more about topics, read https://help.github.com/en/articles/about-topics
- [x] Extension description summarizes in 1-2 sentences what the extension is usable (should be understandable for non-experts)
- [x] Extension URL and revision (scmurl, scmrevision) is correct, consider using a branch name (master, release, ...) instead of a specific git has to avoid re-submitting pull request whenever the extension is updated
- [x] Extension icon URL is correct
- [x] Screenshot URLs (screenshoturls) are correct, contains at least one
- [x] Homepage URL points to valid webpage containing the following:
  - [x] Extension name
  - [x] Short description: 1-2 sentences, which summarizes what the extension is usable for
  - [x] At least one nice, informative image, that illustrates what the extension can do. It may be a screenshot.
  - [x] Description of contained modules: at one sentence for each module
  - [x] Tutorial: step-by-step description of at least the most typical use case, include a few screenshots, provide download links to sample input data set
  - [ ] Publication: link to publication and/or to PubMed reference (if available)
  - [x] License: We suggest you use a permissive license that includes patent and contribution clauses.  This will help protect developers and ensure the code remains freely available.  We suggest you use the [Slicer License](https://github.com/Slicer/Slicer/blob/master/License.txt) or the [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0). Always mention in your README file the license you have chosen.  If you choose a different license, explain why to the extension maintainers. Depending on the license we may not be able to host your work.
Read [here](https://opensource.guide/legal/#which-open-source-license-is-appropriate-for-my-project) to learn more about licenses.

### CODE REVIEW

File: OpenDose4D.s4ext [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+# url to icon (png, size 128x128 pixels)
+iconurl https://gitlab.com/opendose/opendose4d/-/blob/develop/Dosimetry4D/Resources/Icons/Dosimetry4D.png
+
+# Give people an idea what to expect from this code
+#  - Is it just a test or something you stand behind?
+status stable
+
+# One line stating what the module does
+description This module implements the full 3D Dosimetry for molecular radiotherapy on multiple time points.
```
Comment (lassoan) [inline OpenDose4D.s4ext]:
```suggestion
description This module implements full 3D dosimetry for molecular radiotherapy on multiple time points.
```

File: OpenDose4D.s4ext [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+# Firstname1 Lastname1 ([SubOrg1, ]Org1), Firstname2 Lastname2 ([SubOrg2, ]Org2)
+# For example: Jane Roe (Superware), John Doe (Lab1, Nowhere), Joe Bloggs (Noware)
+contributors Alex Vergara Gil (INSERM, France), Janick Rueegger (KSA, Switzerland)
+
+# Match category in the xml description of the module (where it shows up in Modules menu)
+category Nuclear Medicine
+
+# url to icon (png, size 128x128 pixels)
+iconurl https://gitlab.com/opendose/opendose4d/-/blob/develop/Dosimetry4D/Resources/Icons/Dosimetry4D.png
```
Comment (lassoan) [inline OpenDose4D.s4ext]: You need to put image download URL here.

File: OpenDose4D.s4ext [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+# homepage
+homepage http://slicer.org/slicerWiki/index.php/Documentation/Nightly/Extensions/OpenDose4D
+
+# Firstname1 Lastname1 ([SubOrg1, ]Org1), Firstname2 Lastname2 ([SubOrg2, ]Org2)
+# For example: Jane Roe (Superware), John Doe (Lab1, Nowhere), Joe Bloggs (Noware)
+contributors Alex Vergara Gil (INSERM, France), Janick Rueegger (KSA, Switzerland)
+
+# Match category in the xml description of the module (where it shows up in Modules menu)
+category Nuclear Medicine
```
Comment (lassoan) [inline OpenDose4D.s4ext]: Use existing Radiotherapy category. If we have 5+ NM related extensions then we can create a new category for that.

File: OpenDose4D.s4ext [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+# list dependencies
+# - These should be names of other modules that have .s4ext files
+# - The dependencies will be built first
+depends SlicerElastix
+
+# Inner build directory (default is ".")
+build_subdirectory .
+
+# homepage
+homepage http://slicer.org/slicerWiki/index.php/Documentation/Nightly/Extensions/OpenDose4D
```
Comment (lassoan) [inline OpenDose4D.s4ext]: This should be your project homepage - https://gitlab.com/opendose/opendose4d

File: OpenDose4D.s4ext [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#
+# First token of each non-comment line is the keyword and the rest of the line
+# (including spaces) is the value.
+# - the value can be blank
+#
+
+# This is source code manager (i.e. svn)
+scm git
+scmurl git://gitlab.com/opendose/OpenDose4D.git
+scmrevision develop
```
Comment (lassoan) [inline OpenDose4D.s4ext]: Probably you want to use master or release. You can merge commits into master or release branch when changes are ready.

File: OpenDose4D.s4ext [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+# Give people an idea what to expect from this code
+#  - Is it just a test or something you stand behind?
+status stable
+
+# One line stating what the module does
+description This module implements the full 3D Dosimetry for molecular radiotherapy on multiple time points.
+
+# Space separated list of urls
+screenshoturls
```
Comment (lassoan) [inline OpenDose4D.s4ext]: At least a few URLs are required

**Comment by lassoan** [comment]
Thank you for your hard work and excellent contribution. We just now have to ensure that your work is easily discoverable and understandable.

Repository name is expected to Slicer<ExtensionName>. Could you rename your repository to be SlicerOpenDose4D? Also note the capitalization - Opendose is harder to read/interpret than OpenDose.

I would suggest to rework your project page (https://gitlab.com/opendose/opendose4d) to be more user friendly. Currently it mostly contains information that is only interesting for developers and there is no user documentation (tutorials, sample data sets, summary of modules and their features, etc.). Without user documentation, there is no chance that new users cannot find and start using your extension.

Please address the code review comments and complete all the items in the checklist.

File: OpenDose4D.s4ext (hunk 2 above)
Comment (BishopWolf) [reply OpenDose4D.s4ext]: do you mean outside project?

**Comment by BishopWolf** [comment]
Sample datasets are inside in Resources and the user manual is in the project root

**Comment by lassoan** [comment]
> Sample datasets are inside in Resources and the user manual is in the project root.

Provide links to these near the top of https://gitlab.com/opendose/opendose4d/-/blob/develop/README.md

Add a few nice screenshots.

Make sample data easy to download and use. For example, you can add a hidden module that registers custom sample data in SampleData module. This allows users to download data sets by a few clicks in Slicer. See for example here: https://github.com/SlicerRt/Sequences/tree/master/SequenceSampleData

Move developer information (Testing, CI, ...) to the very bottom of the README.md but maybe even into a separate developers.md file, to not overload users with lots of information that is not interesting for them.

> Where do I change the repository name? Is it the project name in gitlab?

Yes.

**Comment by BishopWolf** [comment]
> For example, you can add a hidden module that registers custom sample data in SampleData module.

This is already in place when you hit the Reload and Test button, the whole workflow is tested, including downloading the sample data.

**Comment by lassoan** [comment]
Automated tests are great, Reload&Test is very useful during development - but these are not directly relevant for users.

Providing easy access to sample data is important, as it makes it easier to follow tutorials, try various modules, troubleshoot problems (see if the problem is with the user's data), etc.

**Comment by BishopWolf** [comment]
> Providing easy access to sample data is important, as it makes it easier to follow tutorials, try various modules, troubleshoot problems (see if the problem is with the user's data), etc.

This is harder for me as I can't upload data directly to the midas repository. I think I will rely on the data already provided inside the module as it is already accesible. Also it is explained in the manual.

**Comment by BishopWolf** [comment]
> > Where do I change the repository name? Is it the project name in gitlab?
> 
> Yes.

I have already changed the repository name but the check still fails because of the old name, so what else shall I change?

**Comment by lassoan** [comment]
> I can't upload data directly to the midas repository

You can upload your data anywhere. If you have a set of DICOM files then you can put them all in a zip file and Slicer can download and unzip them.

Can you upload arbitrary set of binary files to gitlab releases? We use github releases to host test/sample data (to avoid blowing up the git repository with lots of binary data): https://github.com/Slicer/SlicerTestingData

**Comment by BishopWolf** [comment]
> You can upload your data anywhere.

I have my sample data in Resources directory, it is part of the gitlab repository

**Comment by lassoan** [comment]
There are a few issues:
- [ ] The module is not packaged correctly. If I install the created package (that you can download from here: https://1drv.ms/u/s!Arm_AFxB9yqHuvEp7Ur3XyaSd9u85w?e=P0TqXb) then the module throws this error:
Traceback (most recent call last):
  File "C:/D/opendose4d_R/_CPack_Packages/win64/ZIP/28862-win-amd64-SlicerOpenDose4D-git-2020-03-20/lib/Slicer-4.11/qt-scripted-modules/Dosimetry4D.py", line 221, in setup
    self.calibration['SPECTSensitivity']['Units'], qt.Qt.MatchFixedString)
KeyError: 'SPECTSensitivity'
- [ ] User manual: Creating user manual in word is not good for many reasons. It is a large binary file. You cannot easily edit (you need to download, load it into word, edit, commit, push; instead of just making changes directly in a web browser), users cannot easily view a word file (it is relatively large, has to be downloaded, needs MS Office to view), and users cannot easily contribute (if a user contributes a small fix then you won't see in the pull request what was changed exactly). Adding a simple markdown files in the repository solves all these issues.
- [ ] The extension package is too large (20MB). What you actually need is 2-3MB python code and maybe a few resource files, but you placed 55MB extra files (sample data files?) in the package. This would increase the burden to the appstore for no good reason. All the large resource files should be downloaded dynamically. SampleData infrastructure can be used for this, it is very robust and the downloaded files are are automatically cached, so you only need to download once.
- [ ] Large binary files are stored in the source code repository. Instead, these large files should be stored in other repositories and files should be downloaded on request. See RSNAVisTutorial.py for an example how to download a DICOM folder and import its content.
- [ ] Sample data sets: The easiest way for users to get started is to add one or more sample data sets in the scene. If the input is typically DICOM then it may be easier if you import the DICOM and save the scene as an MRB file. Could you create such a scene?

**Comment by BishopWolf** [comment]
Thanks for your comments:

1. this is a hidden bug, if installed from gitlab either by cloning or zipping it doesn't show. But it is a bug and will be corrected in the next push
2. Maybe referring to a google doc would be enough, but in this case user will contribute outside gitlab and that is not the intention, I will investigate the markdown way.
3. The package contains sample data (DICOM, 35MB) and a database (SQLITE, 15MB). The DICOM Data is fixed Sample Data that can follow your Sample Data suggestion. The SQLITE database is a critical part of the module and will increase in future, this does not follow the sample data approach as it is a core data in the module, what do you suggest in this case?

**Comment by lassoan** [comment]
15MB database file:
- Storage/revision control: If you change it rarely then it is probably OK to keep it in the source code repository. If you change it every month or so then it is probably better to keep it as a release file (or in a separate repository) to not blow up your source code repository size.
- Installation: If users never modify it and the file compresses well then including it in the install package is probably fine. If users may modify it locally then you probably want to keep a single copy of it in the user's profile folder and only download it if the file does not exist yet.

By the way, I had a look at the content of the database file and it seems extremely inefficient and redundant to use an SQL database for this. You can use VTK data set to store arbitrary set of named arrays (scalars, vectors, strings, etc.) for a set of points or cells. You can also save metadata in field data (to store energy, half life, etc.), so one VTK file can contain all information of an isotope. If interoperability (or the ability of users to easily create/customize isotope files) is more important than performance (it may take 1-2 seconds to parse a long text file) then probably json, csv, or xml format would work much better (probably one file per isotope). These text files also can be compressed really well and since they are text files, you can track their changes in the git repository.

**Comment by BishopWolf** [comment]
Progress Update
* The database has been converted to a set of JSON files
* The Manual has been refactored as MarkDown file.

Pending tasks
* Test the manual fully
* Use the Sample Data upload mechanism for the DICOM data.

**Comment by BishopWolf** [comment]
> * Sample data sets: The easiest way for users to get started is to add one or more sample data sets in the scene. If the input is typically DICOM then it may be easier if you import the DICOM and save the scene as an MRB file. Could you create such a scene?

Is there an example of a testing performed in python using sample data sets? I get the point on uploading the dicom files but I still don't know how to use them from sample data module.

**Comment by lassoan** [comment]
See Sequences extension to see how to register custom sample data sets and Crop volume sequence module for an example of using one of these custom data sets in module tests: https://github.com/SlicerRt/Sequences/blob/3c45c25143f6a2577851e5851747ddbb954b7459/CropVolumeSequence/CropVolumeSequence.py#L268

**Comment by lassoan** [comment]
By the way, very nice progress with the database and manual.

**Comment by lassoan** [comment]
@BishopWolf your extension is in great shape and while there is always room for improvements, there is nothing that should prevent adding the extension to the extension manager. Let me know if there is anything that you would like to still work on before. If not, then we can merge the pull request as is.

**Comment by BishopWolf** [comment]
@lassoan Sorry for the very late response, I was full of work and didn't check the github messages so I totally missed your message until now.

I was previously debugging some instability issues with the display nodes that made slicer to crash. Fortunately that is solved.
We also nearly finished the validation prior to the publication. I expect to send it before october. 

I am currently remaking the database as some users observe some bug in the data uncertainty. After that it can be merged.

**Comment by lassoan** [comment]
I merge this now so we can see any build and packaging errors on the dashboard.

**Comment by BishopWolf** [comment]
> I merge this now so we can see any build and packaging errors on the dashboard.

Is there a reason why the extension doesn't show in the cdash builds? (without errors showing I don't know what happens)

**Comment by jamesobutler** [comment]
Maybe you had checked the dashboard before the extension had been built?

You can use this link to see on the dashboard that there are currently configure errors
http://slicer.cdash.org/index.php?project=SlicerPreview&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=Opendose

**Comment by lassoan** [comment]
It seems that gitlab requires some non-trivial configuration setting for cloning the repository when using git protocol. Probably switching to https protocol will fix the problem.

Not a problem, I'm just curious: why did you choose to use gitlab instead of github?

**Comment by BishopWolf** [comment]
> Not a problem, I'm just curious: why did you choose to use gitlab instead of github?

gitlab has inside CI/CD. It has a gitlab-runner for local tests. it has a full user management and statistics. By the time I started the project github doesn't allowed private repositories.

Currently I highly depend on gitlab-runner to make a switch

**Comment by BishopWolf** [comment]
> t seems that gitlab requires some non-trivial configuration setting for cloning the repository when using git protocol. Probably switching to https protocol will fix the problem.

Usually this is required only for private repositories, which it is not the case. Let me dig in the logs.

**Comment by lassoan** [comment]
Thank you. I heard that github actions have changed the CI/CD landscape and now github has an edge over gitlab. Do you know what additional user control or statistics gitlab provides?

**Comment by BishopWolf** [comment]
> Thank you. I heard that github actions have changed the CI/CD landscape and now github has an edge over gitlab. Do you know what additional user control or statistics gitlab provides?

For private repositories ist has full user management, you assign roles to groups of users. That was really a need when I started. After making the repository public I have no more control than protected branches

but gitlab-runner is now everything to me. I know in github you can run jenkins or circle, but having it running in my own machine allows me to gain time while testing. I put the testing to run and keep programming, when the test ends I make a commit and start a new test. When I want to commit to gitlab then I just squeeze everything.

**Comment by BishopWolf** [comment]
> You can use this link to see on the dashboard that there are currently configure errors

The correct git clone line shall be `git clone git@gitlab.com:opendose/opendose3d.git` and in the error it says it attempted to clone as `git clone git://gitlab.com/opendose/OpenDose3D.git`. 

There is a chance to clone as https as `git clone https://gitlab.com/opendose/OpenDose3D`

**Pull request merged**